DragaliaAPI.Test/Unit/Controllers/DungeonRecordControllerTest.cs

```diff
@@ -0,0 +1,579 @@
+                    this.CreateMissionReward(2),
+                    this.CreateMissionReward(3)
+                }
+            );
+
+        data!.ingame_result_data.reward_record.mission_complete
+            .Should()
+            .BeEquivalentTo(new List<AtgenFirstClearSet>() { this.CreateClearReward() });
+
+        this.mockQuestRepository.VerifyAll();
+        this.mockDungeonService.VerifyAll();
+        this.mockUserDataRepository.VerifyAll();
+        this.mockInventoryRepository.VerifyAll();
+        this.mockUpdateDataService.VerifyAll();
+    }
+
+    // Tests that first clearing all missions but not first clear gives all missions but not first
+    // clear bonus for a total of 20 wyrmite
+    [Fact]
+    public async Task NotFirstClearFullClearGivesAllMissionsButNotFirstClear()
+    {
+        this.mockQuestRepository
+            .Setup(x => x.GetQuests(DeviceAccountId))
+            .Returns(new List<DbQuest>()
+                {
+                    new()
+                    {
+                        DeviceAccountId = DeviceAccountId,
+                        QuestId = questId,
+                        PlayCount = 1,
+                        IsMissionClear1 = false,
+                        IsMissionClear2 = false,
+                        IsMissionClear3 = false
+                    }
+                }.AsQueryable().BuildMock());
+
+        DungeonRecordRecordRequest request = new() { dungeon_key = dungeonKey };
+
+        ActionResult<DragaliaResponse<object>> response = await this.dungeonRecordController.Record(
+            request
+        );
+
+        DungeonRecordRecordData? data = response.GetData<DungeonRecordRecordData>();
+        data.Should().NotBeNull();
+
+        data!.ingame_result_data.reward_record.first_clear_set
+            .Should()
+            .BeEquivalentTo(new List<AtgenFirstClearSet>() { });
+
+        data!.ingame_result_data.reward_record.missions_clear_set
+            .Should()
+            .BeEquivalentTo(
+                new List<AtgenMissionsClearSet>()
+                {
+                    this.CreateMissionReward(1),
+                    this.CreateMissionReward(2),
+                    this.CreateMissionReward(3)
+                }
+            );
+
+        data!.ingame_result_data.reward_record.mission_complete
+            .Should()
+            .BeEquivalentTo(new List<AtgenFirstClearSet>() { this.CreateClearReward() });
+
+        this.mockQuestRepository.VerifyAll();
+        this.mockDungeonService.VerifyAll();
+        this.mockUserDataRepository.VerifyAll();
+        this.mockInventoryRepository.VerifyAll();
+        this.mockUpdateDataService.VerifyAll();
+    }
+
+    // Tests that when previous missions have been completed that the right amount of wyrmite is
+    // given as well as their corresponding mission numbers in missions_clear_set
+    [Fact]
+    public async Task VariousPreviousMissionClearStatusesGiveCorrectMissions()
```

Comment:
When you're doing multiple test cases and you want to assert similarity on all of them, you can do a [Theory] test. This allows you to pass in parameters to the test function and repeat the same code with different values.

Outside of that, [Fact] unit tests should generally only contain one test case in an arrange -> act -> assert pattern. 

Have a look at [TheoryData](https://hamidmosalla.com/2020/04/05/xunit-part-8-using-theorydata-instead-of-memberdata-and-classdata/) in particular -- you may be able to have a list of values like `GetQuests quest`, `CompleteQuests quest` and `expected missions in the response` and then just repeat the same code of setup repository, call method, check response for each case